source/common/quic/envoy_quic_proof_source.cc

```diff
@@ -95,10 +151,40 @@ void EnvoyQuicProofSource::signPayload(
                 std::make_unique<EnvoyQuicProofSourceDetails>(res.filter_chain_.value().get()));
 }
 
-EnvoyQuicProofSource::CertConfigWithFilterChain
-EnvoyQuicProofSource::getTlsCertConfigAndFilterChain(const quic::QuicSocketAddress& server_address,
-                                                     const quic::QuicSocketAddress& client_address,
-                                                     const std::string& hostname) {
+EnvoyQuicProofSource::CertWithFilterChain EnvoyQuicProofSource::getTlsCertAndFilterChain(
+    const quic::QuicSocketAddress& server_address, const quic::QuicSocketAddress& client_address,
+    const std::string& hostname, bool* cert_matched_sni) {
+  ENVOY_LOG(trace, "Getting cert chain for {}", hostname);
+  // TODO(danzh) modify QUICHE to make quic session or ALPN accessible to avoid hard-coded ALPN.
+  Network::ConnectionSocketPtr connection_socket = createServerConnectionSocket(
+      listen_socket_.ioHandle(), server_address, client_address, hostname, "h3");
+  StreamInfo::StreamInfoImpl info(time_source_,
+                                  connection_socket->connectionInfoProviderSharedPtr());
+  const Network::FilterChain* filter_chain =
+      filter_chain_manager_->findFilterChain(*connection_socket, info);
+
+  if (filter_chain == nullptr) {
+    listener_stats_.no_filter_chain_match_.inc();
+    ENVOY_LOG(warn, "No matching filter chain found for handshake.");
+    return {};
+  }
+  ENVOY_LOG(trace, "Got a matching cert chain {}", filter_chain->name());
+
+  auto& transport_socket_factory =
+      dynamic_cast<const QuicServerTransportSocketFactory&>(filter_chain->transportSocketFactory());
```

Comment:
This part is shared between new and legacy code. Maybe extract it into a stand alone method?

Reply:
I was trying to keep it as simple as possible for when the runtime flag is removed and the old code is deleted; I think it's better to have the duplication in case there is a change in the new version; we don't want to automatically apply any changes to the old version as well.

Reply:
The shared part should be the same between new and old versions. That's why I suggest to extract it out. And if you have a handle to the transport socket factory before the code diverge, you can check transport_socket_factory->handle_certs_with_shared_tls_code_ to decide which code to execute.

Reply:
Thanks for making the diff cleaner!